stateresolutionv2.go

```diff
@@ -0,0 +1,565 @@
+			return true, pos
+		}
+		// Loop through the mainline.
+		for pos, mainlineEvent := range r.powerLevelMainline {
+			// Check if the search event matches this event. If it does then the event
+			// is in the mainline.
+			if mainlineEvent.EventID() == searchEvent.EventID() {
+				return true, pos
+			}
+		}
+		// If we've reached this point then the event is not in the mainline.
+		return false, 0
+	}
+
+	// Define our iterator function.
+	var iter func(event Event)
+	iter = func(event Event) {
+		// In much the same way as we do in createPowerLevelMainline, we loop
+		// through the event's auth events, checking that it exists in our supplied
+		// auth event map and finding power level events.
+		for _, authEventID := range event.AuthEventIDs() {
+			// Check that we actually have the auth event in our map - we need this so
+			// that we can look up the event type.
+			if authEvent, ok := r.authEventMap[authEventID]; ok {
+				// Is the event a power level event?
+				if authEvent.Type() == MRoomPowerLevels {
+					// Is the event in the mainline?
+					if isIn, pos := isInMainline(authEvent); isIn {
+						// It is - take a note of the event and position and stop the
+						// iterator from running any further.
+						mainlineEvent = authEvent
+						mainlinePosition = pos
+						// Cache the result so that a future request for this position will
+						// be faster.
+						r.powerLevelMainlinePos[mainlineEvent.EventID()] = mainlinePosition
+						return
+					}
+					// It isn't - increase the step count and then run the iterator again
+					// from the found auth event.
+					steps++
+					iter(authEvent)
+				}
+			}
+		}
+	}
+
+	// Start the iterator with the supplied event.
+	iter(event)
+
+	return
+}
+
+// authAndApplyEvents iterates through the supplied list of events and auths
+// them against the current partial state. If they pass the auth checks then we
+// also apply them on top of the partial state. If they fail auth checks then
+// the event is ignored and dropped. Returns two lists - the first contains the
+// accepted (authed) events and the second contains the rejected events.
+func (r *stateResolverV2) authAndApplyEvents(events []Event) (accepted []Event) {
+	for _, event := range events {
+		// Check if the event is allowed based on the current partial state. If the
+		// event isn't allowed then simply ignore it and process the next one.
+		if err := Allowed(event, r); err != nil {
+			continue
+		}
+		// We've now authed the event.
+		accepted = append(accepted, event)
+		// Apply the newly authed event to the partial state. We need to do this
+		// here so that the next loop will have partial state to auth against.
+		r.applyEvents([]Event{event})
+	}
+	return
+}
+
+// applyEvents applies the events on top of the partial state.
+func (r *stateResolverV2) applyEvents(events []Event) {
+	for _, e := range events {
+		event := e
+		switch event.Type() {
+		case MRoomCreate:
```

Comment:
Is there a reason why we need to store these state events separately, rather than just putting everything in `r.resolvedOthers`?

Reply:
There isn't really a technical reason why we couldn't just store everything in a resolved map at all, no — it's just a bit clearer this way what the functions that implement `AuthEventProvider` do.